vcstool/clients/git.py

```diff
@@ -49,21 +55,82 @@ def _branch(self, all):
     def branch(self, command):
         return self._branch(command.all)
 
-    def compare(self, command):
+    def _unstaged_changes(self):
+        result = self.custom(GitCustomCommand(['diff', '--no-ext-diff', '--quiet']))
+        return result['returncode'] == 1
+
+    def _staged_changes(self):
+        result = self.custom(GitCustomCommand(['diff', '--no-ext-diff', '--cached', '--quiet']))
+        return result['returncode'] == 1
+
+    def _untracked_files(self):
+        result = self.custom(GitCustomCommand(['ls-files', '--others', '--exclude-standard']))
+        return result['output'] != ''
+
+    def _stashes(self):
+        result = self.custom(GitCustomCommand(['rev-parse', '--verify', '--quiet', 'refs/stash']))
+        return result['returncode'] == 0
+
+    def _get_ahead_behind(self, local_branch, remote, remote_branch):
+        """Returns an (int, int) tuple corresponding to the (ahead, behind) commit numbers."""
+        if not remote or not remote_branch:
+            return (0, 0)
+        # remove leading "refs/heads/"
+        prefix = 'refs/heads/'
+        if remote_branch.startswith(prefix):
+            remote_branch = remote_branch[len(prefix):]
+        result = self.custom(GitCustomCommand(['rev-list', '--left-right', '--count',
+                                               f'{local_branch}...{remote}/{remote_branch}']))
+        # If the command failed, just return an error sentinel.
+        if result['returncode']:
+            return (-1, -1)
+        # Parse out the ints and return as a tuple
+        return (int(i) for i in result['output'].strip().split())
+
+    def _get_remote(self, local_branch: str):
+        result =  self.custom(GitCustomCommand(['config', f'branch.{local_branch}.remote']))
+        return result['output'] if result['returncode'] == 0 else ''
 
-        # Get the branches
-        branch_result = self._branch(False)
+    def _get_remote_branch(self, local_branch: str):
+        result =  self.custom(GitCustomCommand(['config', f'branch.{local_branch}.merge']))
+        return result['output'] if result['returncode'] == 0 else ''
 
-        # status_result = self_.status()
+    def _get_tag(self, local_branch : str):
+        result =  self.custom(GitCustomCommand(['describe', '--tags', '--abbrev=0', '--exact-match',
+                                                local_branch]))
+        return result['output'].strip() if result['returncode'] == 0 else ''
+
+    def _get_hash(self, local_branch : str):
+        result = self.custom(GitCustomCommand(['rev-parse', local_branch]))
+        return result['output'].strip() if result['returncode'] == 0 else ''
+
+    def compare(self, command):
+        result = self._branch(False)
+        # If we can't get the local branch, no point in continuing.
```

Comment:
For discussion, how will this `compare` work if the _version_ is a tag or a hash? Do you think that is something we should support?

Reply:
After some futzing about, this is something I came up with. The image below demos a spectrum of possible configurations, but you can note that detached head states are displayed pretty clearly.

Thoughts?

![Selection_005](https://user-images.githubusercontent.com/8943282/210197680-d763990b-8bff-4111-a018-ee725027a1cd.png)

Reply:
I think this make sense and is clear to me. How does this work if the Manifest lists a tag or hash for the version?

Reply:
You can see a few examples of that above, e.g. `ak/ak_mission_executor` -> `some_branch`, `tl/tc_hw_interfaces` -> `4d3c568`